src/rich-text/inputrules.ts

```diff
@@ -15,7 +15,7 @@ const strongRegexAsterisks = /\*\*(\S(?:|.*?\S))\*\*$/;
 // matches: *some text*, but not **text* or * text *
 export const emphasisRegexAsterisk = /\*([^*\s]([^*])*[^*\s]|[^*\s])\*$/;
 // matches: __some text__, but not __ text __
-const strongRegexUnderscores = /__(\S(?:|.*?\S))__$/;
+const strongRegexUnderscores = /(?<!\S)__(\S(?:|.*?\S))__$/;
```

Comment:
Just FYI: For performance reasons it's better to put the lookaround deeper into the regular expression, particularly not as the first thing in the regex.

Basically `/__(?<!\S__)(\S(?:|.*?\S))__$/` will be significantly more performant than `/(?<!\S)__(\S(?:|.*?\S))__$/`. This is because then matching the lookbehind is only evaluated once the `__` matches instead of on *every single character*.

The cheapest thing for a regular expression engine to do is just testing to see if the current character is a fixed character, such as is the current character being a "_". Putting the lookbehind first means that the regex engine does more than double the work on nearly every character (i.e., on every character that's preceded by a `\S` both the preceding character and the current character are tested instead of testing just the current character; and that doesn't even get into it being more expensive to enter and leave a lookbehind and more expensive to compare to a set, which requires comparing against multiple values, than to compare to the single character "_").

What the performance impact is of such a change will depend on the specific regular expression engine, the regular expression, and the text being tested. In general, I've seen it be as little as a 10% improvement to as much as close to 90%+ of the computation. In the real world, we can see 10s of seconds reduction in regex search times on metasmoke by using similar changes in the regex used for the search. 

Obviously, the same performance consideration applies to `emphasisRegexUnderscore` below.

I'm not saying "this needs to change". I'm just indicating something that should be generally considered when creating a regular expression.

Reply:
Thanks for this input. Out of my personal curiosity I have applied the suggestion you made and compare execution timings against our unit test suite. While I believe your point is valid I could not see any relevant improvements in our context at least so we won't retroactively make changes here but it's something we will keep in mind going forward. 
Thanks again.